azure/services/aso/tags.go

```diff
@@ -32,40 +33,32 @@ import (
 const tagsLastAppliedAnnotation = "sigs.k8s.io/cluster-api-provider-azure-last-applied-tags"
 
 // reconcileTags modifies parameters in place to update its tags and its last-applied annotation.
-func reconcileTags(t TagsGetterSetter, existing genruntime.MetaObject, parameters genruntime.MetaObject) error {
+func reconcileTags[T genruntime.MetaObject](t TagsGetterSetter[T], existing T, resourceExists bool, parameters T) error {
+	var existingTags infrav1.Tags
 	lastAppliedTags := map[string]interface{}{}
-	if existing != nil {
+
+	if resourceExists {
 		lastAppliedTagsJSON := existing.GetAnnotations()[tagsLastAppliedAnnotation]
 		if lastAppliedTagsJSON != "" {
 			err := json.Unmarshal([]byte(lastAppliedTagsJSON), &lastAppliedTags)
 			if err != nil {
 				return errors.Wrapf(err, "failed to unmarshal JSON from %s annotation", tagsLastAppliedAnnotation)
 			}
 		}
-	}
 
-	existingTags, err := t.GetActualTags(existing)
-	if err != nil {
-		return errors.Wrapf(err, "failed to get actual tags for %s %s/%s", existing.GetObjectKind().GroupVersionKind(), existing.GetNamespace(), existing.GetName())
+		existingTags = t.GetActualTags(existing)
```

Comment:
There is a little bit of light refactoring in this function not strictly-related to generics to workaround each `GetActualTags` implementation from needing to handle `nil` inputs.